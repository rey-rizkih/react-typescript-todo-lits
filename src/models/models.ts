import { useReducer } from "react";

export interface Todo {
  id: number;
  todo: string;
  isDone: boolean;
}

// type TodoActions = {type:"add" ,payload:string} |{type:"add" ,payload:number }  |{type:"add" ,payload:number}

// const TodoReducer = (state: Todo[], actions: ()) => {}

// const TodoConsumer = () => {
//   const [state, dispatch] = useReducer()
// }
