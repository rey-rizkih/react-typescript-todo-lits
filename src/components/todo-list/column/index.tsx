import { styled } from "@mui/material";
import Box, { BoxProps } from "@mui/material/Box";
import Typography from "@mui/material/Typography";
import { Droppable } from "react-beautiful-dnd";
import type { Column } from "../../../models/models";
import type { ColorVariants } from "../../../styles/theme";

export interface TodoContainerProps {
  variant?: ColorVariants;
  isDraggingOver?: boolean;
}

export interface TodoColumnProps extends BoxProps {
  column: Column & { variant?: ColorVariants };
}

const TodoColumn = ({
  column,
  children,
}: React.PropsWithChildren<TodoColumnProps>) => {
  return (
    <Droppable droppableId={column.id} type="TASK">
      {(provided, snapshot) => (
        <TodoColumnContainer
          ref={provided.innerRef}
          variant={column.variant}
          isDraggingOver={snapshot.isDraggingOver}
          {...provided.droppableProps}
        >
          <Typography variant="h2">{column.title}</Typography>

          {children}

          {provided.placeholder}
        </TodoColumnContainer>
      )}
    </Droppable>
  );
};

const TodoColumnContainer = styled(Box)<TodoContainerProps>(
  ({ theme, variant = "primary", isDraggingOver }) => ({
    borderRadius: "5px",
    display: "flex",
    flexDirection: "column",
    width: "47.5%",
    padding: "15px",

    [theme.breakpoints.down("lg")]: {
      width: "45%",
    },

    [theme.breakpoints.down("md")]: {
      width: "95%",
      marginBottom: "10px",
    },

    background: theme.palette[variant][isDraggingOver ? "dark" : "main"],
  })
);

export default TodoColumn;
