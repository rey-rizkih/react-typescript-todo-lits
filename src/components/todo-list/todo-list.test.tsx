import { cleanup, render } from "@testing-library/react";
import TodoList from ".";
import { TodoColumns, Todos } from "../../models/models";
import { TodoColumnProps } from "./column";

const columns: TodoColumns = {
  active: {
    id: "active",
    title: "Active Tasks",
    rowsId: ["active-1"],
    variant: "secondary",
  },
  completed: {
    id: "completed",
    title: "Completed Tasks",
    rowsId: [],
    variant: "error",
  },
};

const todos: Todos = {
  "active-1": {
    id: "active-1",
    content: "First Task",
    isDone: false,
  },
};

const renderTodoList = (props: Partial<TodoColumnProps> = {}) => {
  render(
    <TodoList
      columns={columns}
      columnOrder={["active", "completed"]}
      todos={todos}
      actions={{
        onEdit: () => {},
        onDelete: () => {},
        onDone: () => {},
      }}
      {...props}
    />
  );
};

const mockTodoColumn = jest.fn();
const mockTodoRow = jest.fn();

jest.mock("./column", () => (props: unknown) => {
  mockTodoColumn(props);
  return <div data-testid="todo-column" {...props}></div>;
});

jest.mock("./rows", () => (props: unknown) => {
  mockTodoRow(props);
  return <div data-testid="todo-row"></div>;
});

afterEach(cleanup);

it("Should render column correctly", async () => {
  renderTodoList();

  // Checking column component props
  expect(mockTodoColumn).toHaveBeenCalledWith(
    expect.objectContaining({
      column: columns.active,
    })
  );

  // Checking column component props
  expect(mockTodoColumn).toHaveBeenCalledWith(
    expect.objectContaining({
      column: columns.completed,
    })
  );
});

it("Should render row correctly", async () => {
  renderTodoList();

  // Checking row component props
  expect(mockTodoRow).toHaveBeenCalledWith(
    expect.objectContaining({
      todos: [todos["active-1"]],
    })
  );

  // Checking row component props
  expect(mockTodoRow).toHaveBeenCalledWith(
    expect.objectContaining({
      todos: [],
    })
  );
});
