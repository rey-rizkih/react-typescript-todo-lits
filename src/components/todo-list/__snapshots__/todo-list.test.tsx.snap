// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot todo list 1`] = `
<div
  className="css-fazsao"
>
  <div
    column={
      Object {
        "id": "active",
        "rowsId": Array [
          "active-1",
        ],
        "title": "Active Tasks",
        "variant": "secondary",
      }
    }
    data-testid="todo-column"
  >
    <div
      data-testid="todo-row"
    />
  </div>
  <div
    column={
      Object {
        "id": "completed",
        "rowsId": Array [],
        "title": "Completed Tasks",
        "variant": "error",
      }
    }
    data-testid="todo-column"
  >
    <div
      data-testid="todo-row"
    />
  </div>
</div>
`;
